File: source/downloads/code/LeetCode/Java/Clone-Graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    // 已经创建过的node,label 映射<label, Node>
    private HashMap<Integer, UndirectedGraphNode> map = new HashMap<>();
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        return clone(node);    
    }
    public UndirectedGraphNode clone(UndirectedGraphNode node){
        if(node==null) return null;
        if(map.containsKey(node.label)) return map.get(node.label);
        
        // 不在map中，说明之前没有给这个节点创建object.
        UndirectedGraphNode newNode = new UndirectedGraphNode(node.label);
        map.put(node.label, newNode);
        // copy neighbors
        for(UndirectedGraphNode neighbor : node.neighbors){
            newNode.neighbors.add(clone(neighbor));
        }
        return newNode;
    }
}

File: downloads/code/LeetCode/Python/Clone-Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

 # Definition for a undirected graph node
# class UndirectedGraphNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution(object):
    def __init__(self):
        # hash_map{label:new_node}
        self.hash_map = {}

    def cloneGraph(self, node):
        """
        :type node: UndirectedGraphNode
        :rtype: UndirectedGraphNode
        """
        # hash_map: {new_node:new_node.label)
        if node == None:
            return None
        if node.label in self.hash_map:
            return self.hash_map[node.label]
        # new node
        root = UndirectedGraphNode(node.label)
        # update hash_map
        self.hash_map[node.label] = root
        for neighbor in node.neighbors:
            root.neighbors.append(self.cloneGraph(neighbor))
        return root      
